api/handlers.go

```diff
@@ -165,3 +171,404 @@ func (api *RecipeAPI) AddRecipeHandler(w http.ResponseWriter, req *http.Request)
+	codelistID := vars["codelist_id"]
+	logD := log.Data{"recipe_id": recipeID, "instance_id": instanceID, "codelist_id": codelistID}
+
+	// Read body
+	b, err := ioutil.ReadAll(req.Body)
+	defer req.Body.Close()
+	if err != nil {
+		log.Event(ctx, "error in reading request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Unmarshal to the shape of Response struct
+	var updates recipe.CodeList
+	err = json.Unmarshal(b, &updates)
+	if err != nil {
+		log.Event(ctx, "error returned from json unmarshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Validating fields of codelist given in request body
+	err = updates.ValidateUpdateCodeList()
+	if err != nil {
+		log.Event(ctx, "bad request error for invalid updates given in request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Find index of specific codelist and interface of codelist in output_instances of the recipe
+	currentRecipe, err := api.dataStore.Backend.GetRecipe(recipeID)
+	if err != nil {
+		log.Event(ctx, "error retrieving specific recipe from mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	instanceIndex := -1
+	codelistIndex := -1
+	for i, instance := range currentRecipe.OutputInstances {
+		if instance.DatasetID == instanceID {
+			instanceIndex = i
+
+			for j, codelist := range instance.CodeLists {
+				if codelist.ID == codelistID {
+					codelistIndex = j
+					break
+				}
+			}
+
+			break
+		}
+	}
+
+	if instanceIndex == -1 {
+		log.Event(ctx, "error retrieving specific instance of recipe from mongo", log.ERROR, logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if codelistIndex == -1 {
+		log.Event(ctx, "error retrieving specific codelist of instance of recipe from mongo", log.ERROR, logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Set the non-updated fields to values of the currentRecipe
+	updates.ID = codelistID
+	if updates.HRef == "" {
+		updates.HRef = currentRecipe.OutputInstances[instanceIndex].CodeLists[codelistIndex].HRef
+	}
+	if updates.Name == "" {
+		updates.Name = currentRecipe.OutputInstances[instanceIndex].CodeLists[codelistIndex].Name
+	}
+	if updates.IsHierarchy == nil {
+		updates.IsHierarchy = currentRecipe.OutputInstances[instanceIndex].CodeLists[codelistIndex].IsHierarchy
+	}
+
+	// Update Recipe to Mongo
+	update := bson.M{"$set": bson.M{"output_instances." + strconv.Itoa(instanceIndex) + ".code_lists." + strconv.Itoa(codelistIndex): updates}}
+	err = api.dataStore.Backend.UpdateRecipe(recipeID, update)
+	if err != nil {
+		log.Event(ctx, "error updating recipe to mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Marshal to output the newly updated recipe
+	updatedRecipe, _ := api.dataStore.Backend.GetRecipe(recipeID)
```

Comment:
Always handle the error please

Reply:
Done in commit "move bson statements to mongo package - mongo package encapsulation"